### DISCUSSION TITLE
`Annotated` in Typer v0.9.0 is broken with `from __future__ import annotations`

### DISCUSSION THREAD

**Discussion by jfcherng** [fastapi/typer#598] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from __future__ import annotations
import typer
from typing_extensions import Annotated


def main(
    name: str,
    lastname: Annotated[str, typer.Option(help="Last name of person to greet.")] = "",
    formal: Annotated[bool, typer.Option(help="Say hi formally.")] = False,
):
    """
    Say hi to NAME, optionally with a --lastname.

    If --formal is used, say hi very formally.
    """
    if formal:
        print(f"Good day Ms. {name} {lastname}.")
    else:
        print(f"Hello {name} {lastname}")


if __name__ == "__main__":
    typer.run(main)
```

### Description

1. Execute the example, which is verbatim taken from https://typer.tiangolo.com/tutorial/options/help/#__tabbed_1_1 except that adding `from __future__ import annotations` to its first line.
2. The help text of args/opts are not shown.

   Actually I think all args for `typer.Option` and `typer.Argument` are neglected.
   
   ![image](https://user-images.githubusercontent.com/6594915/235596668-78b9dc19-72d2-496d-b6cb-7540a57d37a6.png)

### Operating System

Windows

### Typer Version

0.9.0

### Python Version

Python 3.11.3

**Comment by aworkin** [comment]
I'm running into this same issue. `from __future__ import annotations` causes the `annotation` attribute of `inspect.Parameter` objects to be "stringized".

### Where does this happen?

The issue starts in `typer.utils.get_params_from_function()`, at line 109 of typer/typer/utils.py. The signature returned by the `inspect.signature()` call contains `inspect.Parameter` objects whose `annotations` attributes are stringized.

Because these annotations are stringized, the subsequent calls to `typer.utils._split_annotation_from_typer_annotations()` [line 113] with those annotations fail to properly split the Typer annotations, as the check to `typer._typing.get_origin()` [line 98] returns `str` instead of `Annotated`.

### How could this be fixed?

In Python 3.10 and higher, `inspect.signature()` has a new `eval_str` parameter [[doc](https://docs.python.org/3/library/inspect.html#inspect.signature)] as well as the addition of `inspect.get_annotations()` [[doc](https://docs.python.org/3/library/inspect.html#inspect.get_annotations)].

For Python 3.9 and lower, the process is somewhat involved. The annotations how-to [[doc](https://docs.python.org/3/howto/annotations.html#annotations-howto)] suggests emulating the behavior of `inspect.get_annotations()`.

From the how-to:

> In a nutshell, if you wish to evaluate a stringized annotation on an arbitrary object `o`:
    If `o` is a module, use `o.__dict__` as the `globals` when calling [eval()](https://docs.python.org/3/library/functions.html#eval).
    If `o` is a class, use `sys.modules[o.__module__].__dict__` as the `globals`, and `dict(vars(o))` as the `locals`, when calling [eval()](https://docs.python.org/3/library/functions.html#eval).
    If `o` is a wrapped callable using [functools.update_wrapper()](https://docs.python.org/3/library/functools.html#functools.update_wrapper), [functools.wraps()](https://docs.python.org/3/library/functools.html#functools.wraps), or [functools.partial()](https://docs.python.org/3/library/functools.html#functools.partial), iteratively unwrap it by accessing either `o.__wrapped__` or `o.func` as appropriate, until you have found the root unwrapped function.
    If `o` is a callable (but not a class), use `o.__globals__` as the `globals` when calling [eval()](https://docs.python.org/3/library/functions.html#eval).

**Comment by notadoktor** [comment]
Also running into this issue and going slightly mad trying to figure out what was happening.

**Reply by bodograumann** [reply]
Same here. Can we please convert this into a bug issue so it gains more visibility, @tiangolo ?

**Reply by julesbertrand** [reply]
@tiangolo I've run into the same bug using `from __future__ import annotations` in python 3.8 and 3.9

**Reply by bodograumann** [reply]
I just recently learned that `from __future__ import annotations` will be replaced by a new approach, that should hopefully solve all the outstanding issues: https://peps.python.org/pep-0649/

**Comment by svlandeg** [comment]
Related PR: https://github.com/tiangolo/typer/pull/721

**Reply by tiangolo** [reply, marked as answer]
Thanks for the report @jfcherng and for the discussion everyone! :nerd_face: 

This was indeed (probably) solved by #721, available in Typer 0.9.3, released in the next few hours. :rocket:

**Reply by bodograumann** [reply]
Yes it works now :rocket:

**Comment by jfcherng** [comment]
I just found another bug about `typing_extensions.Annotated` and `from __future__ import annotations`.

https://github.com/tiangolo/typer/discussions/802
